File: aoc2024/day22/day22.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <omp.h>
#include <cstdint>

// Constants
constexpr int PRUNE = 0xFFFFFF;
constexpr int NSTEPS = 2000;
constexpr int SEQ_LENGTH = 4;

// Function to transform number (equivalent to f() in original)
uint32_t transform(uint32_t num) {
    num = ((num << 6) ^ num) & PRUNE;
    num = ((num >> 5) ^ num) & PRUNE;
    num = ((num << 11) ^ num) & PRUNE;
    return num;
}

// Shift array left by one and return new value at end
void shift(int arr[SEQ_LENGTH]) {
    for (int i = 0; i < SEQ_LENGTH - 1; i++) {
        arr[i] = arr[i + 1];
    }
}

// Compare two arrays
bool compare(const int arr1[SEQ_LENGTH], const int arr2[SEQ_LENGTH]) {
    for (int i = 0; i < SEQ_LENGTH; i++) {
        if (arr1[i] != arr2[i]) return false;
    }
    return true;
}

// Calculate value for a sequence and initial condition
int calculate_value(const int seq[SEQ_LENGTH], uint32_t ic) {
    int differences[SEQ_LENGTH] = {0};
    int ret = 0;

    for (int i = 0; i < NSTEPS; i++) {
        // Update differences array (equivalent to h() in original)
        shift(differences);
        uint32_t ic_new = transform(ic);
        differences[SEQ_LENGTH-1] = (ic_new % 10) - (ic % 10);
        ic = ic_new;

        // Check for sequence match after we have enough values
        if (i >= 3 && compare(differences, seq)) {
            ret += (ic % 10);
            break;
        }
    }

    return ret;
}

int main() {
    // Read input file
    std::vector<uint32_t> initial_conditions;
    std::ifstream infile("input.txt");
    uint32_t num;

    while (infile >> num) {
        initial_conditions.push_back(num);
    }
    std::cout<<initial_conditions.size()<<std::endl;

    // Set number of threads
    omp_set_num_threads(20);

    int max_result = 0;
    int progress_counter = 0;
    const int total_iterations = 19 * 19;

    // Parallel search through all possibilities
    #pragma omp parallel for collapse(4) schedule(dynamic) reduction(max:max_result)
    for (int a = -9; a <= 9; a++) {
        for (int b = -9; b <= 9; b++) {
            for (int c = -9; c <= 9; c++) {
                for (int d = -9; d <= 9; d++) {
                    int seq[SEQ_LENGTH] = {a, b, c, d};
                    int total = 0;

                    // Calculate sum for all initial conditions
                    for (uint32_t ic : initial_conditions) {
                        total += calculate_value(seq, ic);
                    }

                    max_result = std::max(max_result, total);
                }
            }
        }
        /*
        // Progress indication (outside parallel region)
        #pragma omp critical
        {
            progress_counter++;
            std::cout << (progress_counter * 100.0) / total_iterations << "% complete\r" << std::flush;
        }*/
    }

    std::cout << "\nFinal result: " << max_result << std::endl;
    return 0;
}

File: aoc2024/day22/day22.js
const fs = require('fs');

let readFileAsString=function(fname){
    return fs.readFileSync(__dirname+"/"+fname,{encoding:'utf8',flag:'r'});
};
const testinput=readFileAsString("testinput.txt");
//const testinput2=readFileAsString("testinput2.txt");
const input=readFileAsString("input.txt");

let Differences= arr=>arr.slice(1).map((b,a)=>b-arr[a]);
let Sum=arr=>arr.reduce((a,b)=>a+b,0);
let Last=arr=>arr[arr.length-1];
let First=arr=>arr[0];

let day22a=function(input){
    console.log("=========== day 22 part 1 ==========");
    let ics=input.split('\n').map( x=> x.trim()).filter(line => line.length>0).map(x => +x);

    //0xFFFFFF
    let f=function(num){
        const prune=16777216-1;
        num=((num<<6)^num)&prune;
        num=((num>>5)^num)&prune;
        num=(((num<<11)^num)&prune);
        return num;
    }
    let nest=function(f,num,ntimes){
        for(let n=0;n<ntimes;n++){
            num=f(num);
        }
        return num;
    }
    let ret=0;
    for(let ic of ics){
        //console.log(ic);
        ret+=nest(f,ic>>0,2000);
    }

    console.log(ret);
};

let day22b=function(input){
    console.log("=========== day 22 part 2 ==========");
    let ics=input.split('\n').map( x=> x.trim()).filter(line => line.length>0).map(x => +x);

    //0xFFFFFF
    let f=function(num){
        const prune=16777216-1;
        num=((num<<6)^num)&prune;
        num=((num>>5)^num)&prune;
        num=(((num<<11)^num)&prune);
        return num;
    }
    let nest=function(f,num,ntimes){
        for(let n=0;n<ntimes;n++){
            num=f(num);
        }
        return num;
    }
    let value=function(seq,ic){
        let differences=[0,0,0,0];
        const nsteps=2000;
        let shift=function(list){
            for(let i=0;i<list.length-1;i++)
                list[i]=list[i+1];
            return list;
        }
        let h=function(list,ic){
            shift(list);
            let icnew=f(ic);
            list[list.length-1]=(icnew%10)-(ic%10);
            return [list,icnew];
        }
        let cmp=function(list1,list2){
            for(let i=0;i<list1.length;i++){
                if(list1[i]!=list2[i])
                    return false;
            }
            return true;
        }
        let ret=0;
        for(let i=0;i<nsteps;i++){
            [differences,ic]=h(differences,ic);
            if(i>=3 && cmp(differences,seq))
                ret+=(ic%10);
        }
        return ret;
    };
    let ret=0;

    for(let a=-9;a<=9;a++){
    for(let b=-9;b<=9;b++){
    for(let c=-9;c<=9;c++){
    for(let d=-9;d<=9;d++){
        let total=0;
        for(let ic of ics){
            total+=value([a,b,c,d],ic);
        }
        if(total>ret)
            ret=total;
    }
    }
        console.log(((19*(a+9)+(b+9))/(19*19))+ " percent done");
    }
    }
    console.log(ret);
};

//day22a(testinput);
//day22a(input);
//day22b(testinput);
day22b(input);
